pkg/storage/tsdb/tombstones.go

```diff
@@ -0,0 +1,360 @@
+package tsdb
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/thanos-io/thanos/pkg/objstore"
+
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+	util_log "github.com/cortexproject/cortex/pkg/util/log"
+)
+
+type BlockDeleteRequestState string
+
+const (
+	StatePending   BlockDeleteRequestState = "pending"
+	StateProcessed BlockDeleteRequestState = "processed"
+	StateCancelled BlockDeleteRequestState = "deleted"
+)
+
+// Relative to user-specific prefix.
+const TombstonePath = "tombstones/"
+
+var (
+	ErrTombstoneAlreadyExists      = errors.New("The deletion tombstone with the same request information already exists")
+	ErrInvalidDeletionRequestState = errors.New("Deletion request filename extension indicating the state is invalid")
+	ErrTombstoneNotFound           = errors.New("Tombstone file not found in the object store")
+	ErrTombstoneDecode             = errors.New("Unable to read tombstone contents from file")
+	AllDeletionStates              = []BlockDeleteRequestState{StatePending, StateProcessed, StateCancelled}
+)
+
+type Tombstone struct {
+	RequestCreatedAt int64                   `json:"request_created_at"`
+	StateCreatedAt   int64                   `json:"state_created_at"`
+	RequestID        string                  `json:"request_id"`
+	StartTime        int64                   `json:"start_time"`
+	EndTime          int64                   `json:"end_time"`
+	Selectors        []string                `json:"selectors"`
+	Matchers         []*labels.Matcher       `json:"-"`
+	UserID           string                  `json:"user_id"`
+	State            BlockDeleteRequestState `json:"state"`
+}
+
+func NewTombstone(userID string, requestTime int64, stateTime int64, startTime int64, endTime int64, selectors []string, requestID string, state BlockDeleteRequestState) *Tombstone {
+	return &Tombstone{
+		RequestCreatedAt: requestTime,
+		StateCreatedAt:   stateTime,
+		StartTime:        startTime,
+		EndTime:          endTime,
+		Selectors:        selectors,
+		UserID:           userID,
+		RequestID:        requestID,
+		State:            state,
+	}
+}
+
+// TombstoneManager is responsible for reading and writing tombstone files to the bucket.
+type TombstoneManager struct {
+	bkt    objstore.InstrumentedBucket
+	logger log.Logger
+}
+
+func NewTombstoneManager(
+	bkt objstore.Bucket,
+	userID string,
+	cfgProvider bucket.TenantConfigProvider,
+	logger log.Logger) *TombstoneManager {
+
+	return &TombstoneManager{
+		bkt:    bucket.NewUserBucketClient(userID, bkt, cfgProvider),
+		logger: util_log.WithUserID(userID, logger),
+	}
+}
+
+// Uploads a tombstone file to object sotre
+func (m *TombstoneManager) WriteTombstoneFile(ctx context.Context, tombstone *Tombstone) error {
```

Comment:
Is it a "file"?

Reply:
It does write a file to the object store which is why I originally named it that. However putting "file" might be redundant, just the same way when writing the block deletion marker, it doesn't say "marker file". I've renamed it to just `WriteTombstone`, cause I think the file part is implied. Do you see any potential concerns with this?